### PR TITLE
Initial architecture doc

### PR DESCRIPTION
**Pull request by ostromart**

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+The operator code is divided roughly into five areas:
+
+1. [New API](The_new_API) and related infrastructure. The new API is expressed as a
+[proto](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto) and
+compiled to [Go structs](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.pb.go). The new
+API has pass-through fields to the Helm values.yaml API, but these are additionally validated through a
+[schema](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/values_types.go).
+1. [Controller](K8s_controller) code. The code comprises the K8s listener, webhook and logic for reconciling the cluster
+to an IstioControlPlane CR.
+1. [Manifest creation](Manifest_creation) code. Manifest rendering is done by overlaying user settings on top of a
```
Comment (ayj) [inline ARCHITECTURE.md]: "Render Kubernetes manifests for the Istio Control Plane deployments from the user provided IstioControlPlane API specification. This involves applying the user's settings onto the selected base profile before Helm template rendering. Afterwards any optional user provided overlay patches are applied to the rendered manifests".

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+selected profile and passing the resulting values to a renderer in the Helm library to create manifests. Further
+overlays on the created manifests may be applied if they are defined in the user settings.
+1. [CLI](CLI) code. CLI code shares the IstioControlPlane API with the controller, but allows manifests to be generated
```
Comment (ayj) [inline ARCHITECTURE.md]: We also probably want an operational status/check area. This would be used by the operator for self-healing and CRD status. It could also be exposed through the CLI, e.g. `istioctl check`.

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+and optionally applied from the command line without the need to run a privileged controller in the cluster.
+1. [Migration tools](Migration_tools). The migration tools are intended to automate configuration migration from Helm to
+the operator.
+
+The operator code uses the new Helm charts in the [istio/installer](https://github.com/istio/installer) repo. It is not
+compatible with the older charts in [istio/istio](https://github.com/istio/istio/tree/master/install/kubernetes/helm).
+See the istio/installer repo for details about the new charts and why they were created. Briefly, the new charts
+are intended to support more realistic production deployments of Istio and hitless upgrades with canarying.
```
Comment (ayj) [inline ARCHITECTURE.md]: I'm not sure "realistic" adds much here. Maybe "... intended to support production ready Istio deployments following industry best practices, e.g. canary'd rollouts"

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+## Terminology
+
+Throughout the document, the following terms are used:
+
+- New API: The API directly defined in the
```
Comment (ayj) [inline ARCHITECTURE.md]: "New API" won't age well :) Can we use `IstioControlPlane` instead? We can add a version qualifier if necessary to different the API maturity.

File: ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    citadel:
+      enabled: false
+```
+
+will enable all components of the Security feature except citadel.
+
+These rules are expressed in code in the
+[name](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/name/name.go#L38) package.
+
+### K8s settings
```
Comment (ayj) [inline ARCHITECTURE.md]: AFAICT, all of these settings affect the deployment. Are there any use cases for exposing non-deployment related k8s settings here?

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+Both the new API and old Helm API are validated. The new API is validated through a table of validation rules in
+[pkg/validate/validate.go](https://github.com/istio/operator/blob/master/pkg/validate/validate.go). These rules
+refer to the Go struct path schema and hence have names with a capitalized first letter.
+The old Helm values.yaml API is validated in
+[validate_values.go](https://github.com/istio/operator/blob/master/pkg/validate/validate_values.go)
+and refer to the values.yaml data paths. Hence, these rules have names with a lower case first letter.
+
+## K8s controller
+
+WIP.
```
Comment (sdake) [inline ARCHITECTURE.md]: i'd recommend TODO(ostromart) as these re easier to grep in the code base.

File: ARCHITECTURE.md [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+## Manifest creation
+
+Manifest rendering is a multi-step process, shown in the figure below.
```
Comment (sdake) [inline ARCHITECTURE.md]: please use `SVG` for all images, not png.

File: ARCHITECTURE.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+and the old Helm values (Istio behavior configuration).
+1. The new API fields defined in the user CR override any values defined in the configuration profile CR.  The
+resulting CR is converted to Helm values.yaml format and passed to the next step.
+1. Part of the configuration profile contains settings in the old API values.yaml schema format. User overrides of
+these fields are applied and merged with the output of this step. The result of this step is a merge of configuration
+profile defaults and user overlays, all expressed in Helm values.yaml format. This final values.yaml configuration
+is passed to the Helm rendering library and used to render the charts. The rendered manifests are passed to the next
+step.
+1. Overlays in the user CR are applied to the rendered manifests. No values are ever defined in configuration profile
+CRs at this layer, so no merge is performed in this step.
```
Comment (sdake) [inline ARCHITECTURE.md]: this whole section refers to new api and old api.  Instead of referring to them by age refer to them by some other form.  new and old are hard to parse (what functioanlity is in the old what is in the new).  This is a naming problem :)

File: ARCHITECTURE.md [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+The CLI `iop` command is implemented in the [cmd/iop](https://github.com/istio/operator/blob/master/cmd/iop/)
+subdirectory as a Cobra command with the following subcommands:
+
+- [manifest](https://github.com/istio/operator/blob/master/cmd/iop/manifest.go): renders a manifest and outputs to console or files.
+- [install](https://github.com/istio/operator/blob/master/cmd/iop/install.go): renders and manifest and applies it to a cluster.
+- [dump-profile](https://github.com/istio/operator/blob/master/cmd/iop/dump.go): dumps the default values for a selected profile.
+
+## Migration tools
+
+WIP.
```
Comment (sdake) [inline ARCHITECTURE.md]: TODO(ostromart) plese.

File: ARCHITECTURE.md [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+The available features and components are as follows:
+
+| Feature | Components |
+|---------|------------|
+Base | CRDs
+Traffic Management | Pilot
+Policy | Policy
+Telemetry | Telemetry
+Security | Citadel
+| | Node agent
```
Comment (sdake) [inline ARCHITECTURE.md]: are these not implemented?  if so, ight add a third field which is there state (Not implemented, alpha, beta, stable).

File: ARCHITECTURE.md [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+    nodeAgent:
+      namespace: istio-security-nodeagent
+policy:
+  components:
+    policy:
+```
+
+the resulting namespaces will be:
+
+|Component|Namespace|
```
Comment (sdake) [inline ARCHITECTURE.md]: md requirees spaces on each side of block in a header please change to `| Component | Namesapce |`

File: ARCHITECTURE.md [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+```yaml
+security:
+  enabled: true
+  components:
+    citadel:
+      enabled: false
+```
+
+will enable all components of the Security feature except citadel.
```
Comment (sdake) [inline ARCHITECTURE.md]: lowercase "Security".  I am not sure if citadel should be capitalized - might ask @rcaballeromx or @geeknoid.

File: ARCHITECTURE.md [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+These rules are expressed in code in the
+[name](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/name/name.go#L38) package.
+
+### K8s settings
+
+Rather than defining selective mappings from parameters to fields in K8s resources, the new API contains a consistent
+K8s block for each Istio component. The available K8s settings are defined in
+[KubernetesResourcesSpec](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto#L411):
```
Comment (sdake) [inline ARCHITECTURE.md]: is this line number persistent?

**Comment by geeknoid** [comment]
Sorry to keep bugging about this kind of stuff but...

Any reason why this is not a doc in our team drive? We don't keep design material in our source repos in general.

File: ARCHITECTURE.md (hunk 1 above)
Comment (ostromart) [reply ARCHITECTURE.md]: I reworked this a bit to be clearer and avoid passive voice. lmkwyt

File: ARCHITECTURE.md (hunk 3 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 4 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Good point, done. I've changed references to new and old APIs throughout the doc.

File: ARCHITECTURE.md (hunk 10 above)
Comment (ostromart) [reply ARCHITECTURE.md]: They are, I think the table layout was just not clear. Updated.

File: ARCHITECTURE.md (hunk 11 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 12 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 5 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Yes, we have HPA settings.

File: ARCHITECTURE.md (hunk 13 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Yes, because it's at a sha.

File: ARCHITECTURE.md (hunk 7 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 6 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 8 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 9 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

**Comment by ostromart** [comment]
@geeknoid, the purpose of this text is to explain the code for developers wishing to contribute rather than being a full design doc (which we do have in docs). I've changed the title to "code overview" to reflect that and added a ptr to the full design doc.

File: ARCHITECTURE.md [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+1. [CLI](CLI) code. CLI code shares the `IstioControlPlaneSpec` API with
+the controller, but allows manifests to be generated and optionally applied from the command line without the need to
+run a privileged controller in the cluster. 
+1. [Migration tools](Migration_tools). The migration tools are intended to
+automate configuration migration from Helm to the operator.
+
+The operator code uses the new Helm charts in the [istio/installer](https://github.com/istio/installer) repo. It is not
+compatible with the older charts in [istio/istio](https://github.com/istio/istio/tree/master/install/kubernetes/helm).
+See the istio/installer repo for details about the new charts and why they were created. Briefly, the new charts
+are intended to support production ready deployments of Istio that follow best practises like canarying for upgrade.
```
Comment (rlenglet) [inline ARCHITECTURE.md]: practises -> practices

File: ARCHITECTURE.md [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+
+## Terminology
+
+Throughout the document, the following terms are used:
+
+- `IstioControlPlaneSpec`: The API directly defined in the
+[IstioControlPlaneSpec proto](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto),
+including feature and component groupings, namespaces and enablement, and per-component K8s settings. The term new API
```
Comment (rlenglet) [inline ARCHITECTURE.md]: new API -> "new API" or _new API_

File: ARCHITECTURE.md (hunk 8 above)
Comment (rlenglet) [reply ARCHITECTURE.md]: Please use "IstioControlPlaneSpec API" consistently instead of "new API". There are still a few occurrences of "new API" in the doc.

File: ARCHITECTURE.md [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+
+### Features and components
+
+The operator has a very similar structure to istio/installer: components are grouped into features.
+`IstioControlPlaneSpec` defines functional settings at the feature level. Functional settings are those that performs some
+function in the Istio control plane without necessarily being tied to any one component that runs in a Deployment.
+Component settings are those that necessarily refer to a particular Deployment or Service. For example, the number
+of Pilot replicas is clearly a component setting, because it refers to a component which is a Deployment in the
+cluster. Most K8s platform settings are necessarily component settings.
+The available features and the components that comprise it are as follows:
```
Comment (rlenglet) [inline ARCHITECTURE.md]: comprise it -> comprise each feature

File: ARCHITECTURE.md [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+
+The `IstioControlPlaneSpec` API is intended to replace the installation and K8s parts of Helm values.yaml.
+
+### Features and components
+
+The operator has a very similar structure to istio/installer: components are grouped into features.
+`IstioControlPlaneSpec` defines functional settings at the feature level. Functional settings are those that performs some
+function in the Istio control plane without necessarily being tied to any one component that runs in a Deployment.
+Component settings are those that necessarily refer to a particular Deployment or Service. For example, the number
+of Pilot replicas is clearly a component setting, because it refers to a component which is a Deployment in the
```
Comment (rlenglet) [inline ARCHITECTURE.md]: Remove "clearly".

File: ARCHITECTURE.md [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+nodeAgent | istio-security-nodeagent
+
+These rules are expressed in code in the
+[name](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/name/name.go#L38) package.
+
+### Enablement
+
+Features and components can be individually or collectively enabled or disabled. If a feature is disabled, all of its
+components are disabled, regardless of their component level enablement. If a feature is enabled, all of its components
+are enabled, unless they individually disabled. For example:
```
Comment (rlenglet) [inline ARCHITECTURE.md]: unless they -> unless they are

File: ARCHITECTURE.md [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+Security | Node agent
+Security | Cert manager
+Configuration management | Galley
+Gateways | Ingress gateway
+Gateways | Egress gateway
+AutoInjection | Sidecar injector
+
+Features and components are defined in the
+[name](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/name/name.go#L38) package.
+
```
Comment (rlenglet) [inline ARCHITECTURE.md]: One question I have at this point is whether the operator manages dependencies between features and/or between components. E.g. does it disable Pilot automatically if CRDs is not enabled?
There is no mention of dependency management in the rest of the doc, so I assume the answer is "no."
I think it would be great to add a line here, or in the Enablement section to clarify that explicitly.

File: ARCHITECTURE.md [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+citadel | istio-security
+nodeAgent | istio-security-nodeagent
+
+These rules are expressed in code in the
+[name](https://github.com/istio/operator/blob/905dd84e868a0b88c08d95b7ccf14d085d9a6f6b/pkg/name/name.go#L38) package.
+
+### Enablement
+
+Features and components can be individually or collectively enabled or disabled. If a feature is disabled, all of its
+components are disabled, regardless of their component level enablement. If a feature is enabled, all of its components
```
Comment (rlenglet) [inline ARCHITECTURE.md]: component level -> component-level

File: ARCHITECTURE.md (hunk 8 above)
Comment (rlenglet) [reply ARCHITECTURE.md]: Also, there are still occurrences of "old" in the doc. To me, this implies that this functionality will get deprecated at some point. Is that correct? What is the plan to replace Helm values.yaml in the operator? It seems like it's still an important part of the manifest generation. If this won't get deprecated, don't call it "old".

File: ARCHITECTURE.md [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+and refer to the values.yaml data paths. Hence, these rules have names with a lower case first letter.
+
+## K8s controller
+
+TODO(rcernich).
+
+## Manifest creation
+
+Manifest rendering is a multi-step process, shown in the figure below. ![rendering
+process](images/operator_render_flow.svg) The example in the figure shows the rendering being triggered by a CLI `iop`
```
Comment (rlenglet) [inline ARCHITECTURE.md]: The diagram mentions "remote release URL," so I assume that the operator has the ability to download Helm charts? How does it work? Where are the charts stored? Are they periodically refreshed? How? Is the operator's filesystem read-only?

If you don't intend to document this feature, remove it from the diagram.

File: ARCHITECTURE.md (hunk 14 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 15 above)
Comment (ostromart) [reply ARCHITECTURE.md]: I've removed all references to the old and new API.

File: ARCHITECTURE.md (hunk 17 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 16 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 19 above)
Comment (ostromart) [reply ARCHITECTURE.md]: The answer is it only validates dependencies but does not automatically mutate configs. I've added some wording to that effect in the Validations section. 
For mutating the config, I think it's something we can consider for the future. I think for now it's probably clear enough that we don't do such things without making an explicit statement about it.

File: ARCHITECTURE.md (hunk 21 above)
Comment (ostromart) [reply ARCHITECTURE.md]: The feature is still not fully baked but will be there in the release. I've put a TODO here for now.

**Comment by sdake** [comment]
@ostromart reference https://github.com/istio/operator/pull/90 for an example of what prow checks should look like.  The way to get the prow checks and circleci checks in sync with this PR is to force push after a master rebase e.g.:
```
git fetch upstream
git rebase upstream/master
git push -f origin/branchname
```

A pictorial representation of what the checks should look like:

part 1:
Screen Shot 2019-07-16 at 5 16 55 AM

part 2:
Screen Shot 2019-07-16 at 5 17 08 AM

Cheers
-steve

**Comment by sdake** [comment]
> Sorry to keep bugging about this kind of stuff but...
> 
> Any reason why this is not a doc in our team drive? We don't keep design material in our source repos in general.

@geeknoid 
What would make sense here is to have the `CONTRIBUTING.md` guide as normal, and then (at the choice of the repository maintainers) have a `CONTRIBUTING_DETAILS.md` file or simply `DETAILS.md` file.  This doc isn't quite that, but its pretty close and I believe is the intent of @ostromart.

Cheers
-steve

**Comment by sdake** [comment]
**NB** I am going to keep this PR open to reduce the amount of churn I do while testing the CI jobs in prow.

1. Plase do not merge
2. please ignore this PR unless you want to follow along

File: ARCHITECTURE.md (hunk 18 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.

File: ARCHITECTURE.md (hunk 20 above)
Comment (ostromart) [reply ARCHITECTURE.md]: Done.
